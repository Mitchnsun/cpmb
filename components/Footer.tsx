import Image from "next/image";
import Link from "next/link";

import artists from "@/assets/contents/artists.json";
import Heading from "@/components/Heading";

const Footer = () => {
  const artistNames = Object.keys(artists);

  return (
    <footer className="font-noto">
      <section className="container mx-auto flex items-center justify-around gap-6 py-4 text-center">
        <a
          href="https://www.veran-piano.com/"
          aria-label="Visitez le site de Veran Piano - nouvelle fenêtre"
          target="_blank"
          rel="noopener noreferrer"
          className="basis-1/3"
        >
          <Image
            src="/logo-veran-pianos.png"
            alt="Logo Veran Piano"
            width={200}
            height={120}
            sizes="200px"
            className="m-auto h-auto"
          />
        </a>
        <a
          href="https://www.gaillard.fr/"
          aria-label="Visitez le site de la Ville de Gaillard - nouvelle fenêtre"
          target="_blank"
          rel="noopener noreferrer"
          className="basis-1/3"
        >
          <Image
            src="/logo-gaillard.png"
            alt="Logo Ville de Gaillard"
            width={200}
            height={120}
            sizes="200px"
            className="m-auto h-auto"
          />
        </a>
        <a
          href="https://www.hautesavoie.fr/"
          aria-label="Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre"
          target="_blank"
          rel="noopener noreferrer"
          className="basis-1/3"
        >
          <Image
            src="/haute-savoie.svg"
            alt="Logo Haute-Savoie"
            width={90}
            height={55}
            sizes="90px"
            className="m-auto h-auto"
          />
        </a>
      </section>
      <div className="bg-gray-800 px-4">
        <div className="container mx-auto flex flex-col items-start gap-4 py-4 lg:flex-row lg:gap-24">
          <div>
            <Image
              src="/CPMB-logo-blanc.png"
              alt="Logo Chœur des Pays du Mont-Blanc"
              className="h-auto shrink-0"
              sizes="160px"
              width={160}
              height={55}
            />
            <p className="mt-4 text-sm text-gray-300">Partager la passion de la musique chorale au cœur des Alpes.</p>
          </div>
          <section>
            <Heading hLevel={4} variant={3} className="mb-2 text-white">
              Nos Partenaires
            </Heading>
            <ul className="space-y-1 text-sm text-gray-300">
              <li>
                <a
                  href="https://www.veran-piano.com/"
                  aria-label="Visitez le site de Veran Piano - nouvelle fenêtre"
                  className="underline-offset-2 hover:text-white focus-visible:underline"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  Veran Piano
                </a>
              </li>
              <li>
                <a
                  href="https://www.gaillard.fr/"
                  aria-label="Visitez le site de la Ville de Gaillard - nouvelle fenêtre"
                  className="underline-offset-2 hover:text-white focus-visible:underline"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  Ville de Gaillard
                </a>
              </li>
              <li>
                <a
                  href="https://www.hautesavoie.fr/"
                  aria-label="Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre"
                  className="underline-offset-2 hover:text-white focus-visible:underline"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  Département de la Haute-Savoie
                </a>
              </li>
            </ul>
          </section>
          <section>
            <Heading hLevel={4} variant={3} className="mb-2 text-white">
              Plan du site
            </Heading>
            <ul className="space-y-1 text-sm text-gray-300">
              <li>
                <Link href="/" className="underline-offset-2 hover:text-white focus-visible:underline">
                  Page d&apos;accueil
                </Link>
              </li>
              <li>
                <Link href="/presentation" className="underline-offset-2 hover:text-white focus-visible:underline">
                  Présentation
                </Link>
              </li>
              <li>
                <Link href="/nos-concerts" className="underline-offset-2 hover:text-white focus-visible:underline">
                  Nos concerts
                </Link>
              </li>
              <li>
                <Link href="/presse" className="underline-offset-2 hover:text-white focus-visible:underline">
                  Presse
                </Link>
              </li>
              <li>
                <Link href="/mentions-legales" className="underline-offset-2 hover:text-white focus-visible:underline">
                  Mentions légales
                </Link>
              </li>
            </ul>
          </section>
          <section>
            <Heading hLevel={4} variant={3} className="mb-2 text-white">
              Artistes
            </Heading>
            <ul className="space-y-1 text-sm text-gray-300">
              {artistNames.map((artistKey) => (
                <li key={artistKey}>
                  <Link
                    href={`/presentation/${artistKey}`}
                    className="underline-offset-2 hover:text-white focus-visible:underline"
                  >
                    {artists[artistKey as keyof typeof artists].name}
                  </Link>
                </li>
              ))}
            </ul>
          </section>
        </div>
        <p className="container mx-auto border-t border-gray-700 py-4 text-center text-sm text-gray-300">
          &copy; {new Date().getFullYear()} Chœur des Pays du Mont-Blanc. Tous droits réservés - Conçu et développé
          par&nbsp;
          <a
            href="https://www.gocosmic.dev/fr"
            target="_blank"
            rel="noopener noreferrer"
            aria-label="Visitez le site de Go Cosmic - nouvelle fenêtre"
            className="text-sky-400 underline decoration-sky-400 underline-offset-2 hover:text-sky-300 focus-visible:ring-2 focus-visible:ring-sky-400 focus-visible:ring-offset-2 focus-visible:ring-offset-gray-800 focus-visible:outline-none"
          >
            Go Cosmic
          </a>
        </p>
      </div>
    </footer>
  );
};

export default Footer;
