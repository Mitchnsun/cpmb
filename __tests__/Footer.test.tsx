import { render, screen } from "@testing-library/react";
import type { ImageProps } from "next/image";
import Footer from "@/components/Footer";

// Mock Next.js components
vi.mock("next/image", () => ({
  default: ({ src, alt, width, height, className }: ImageProps) => (
    // eslint-disable-next-line @next/next/no-img-element
    <img src={src} alt={alt} width={width} height={height} className={className} />
  ),
}));

describe("Footer", () => {
  beforeEach(() => {
    // Mock Date.getFullYear() to return a consistent year for testing
    vi.useFakeTimers();
    vi.setSystemTime(new Date("2024-01-01"));
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it("should render the footer element", () => {
    render(<Footer />);

    expect(screen.getByRole("contentinfo")).toBeInTheDocument();
  });

  describe("Partner logos section", () => {
    it("should render all partner logos with correct attributes", () => {
      render(<Footer />);

      // Veran Piano
      const veranLogo = screen.getByAltText("Logo Veran Piano");
      expect(veranLogo).toBeInTheDocument();
      expect(veranLogo).toHaveAttribute("src", "/logo-veran-pianos.png");
      expect(veranLogo).toHaveAttribute("width", "200");
      expect(veranLogo).toHaveAttribute("height", "120");

      // Gaillard
      const gaillardLogo = screen.getByAltText("Logo Ville de Gaillard");
      expect(gaillardLogo).toBeInTheDocument();
      expect(gaillardLogo).toHaveAttribute("src", "/logo-gaillard.png");
      expect(gaillardLogo).toHaveAttribute("width", "200");
      expect(gaillardLogo).toHaveAttribute("height", "120");

      // Haute-Savoie
      const hauteSavoieLogo = screen.getByAltText("Logo Haute-Savoie");
      expect(hauteSavoieLogo).toBeInTheDocument();
      expect(hauteSavoieLogo).toHaveAttribute("src", "/haute-savoie.svg");
      expect(hauteSavoieLogo).toHaveAttribute("width", "90");
      expect(hauteSavoieLogo).toHaveAttribute("height", "55");
    });

    it("should have partner links with correct URLs and accessibility attributes", () => {
      render(<Footer />);

      // Veran Piano link
      const veranLink = screen.getByLabelText("Visitez le site de Veran Piano - nouvelle fenêtre");
      expect(veranLink).toHaveAttribute("href", "https://www.veran-piano.com/");
      expect(veranLink).toHaveAttribute("target", "_blank");
      expect(veranLink).toHaveAttribute("rel", "noopener noreferrer");

      // Gaillard link
      const gaillardLink = screen.getByLabelText("Visitez le site de la Ville de Gaillard - nouvelle fenêtre");
      expect(gaillardLink).toHaveAttribute("href", "https://www.gaillard.fr/");
      expect(gaillardLink).toHaveAttribute("target", "_blank");
      expect(gaillardLink).toHaveAttribute("rel", "noopener noreferrer");

      // Haute-Savoie link
      const hauteSavoieLink = screen.getByLabelText(
        "Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre"
      );
      expect(hauteSavoieLink).toHaveAttribute("href", "https://www.hautesavoie.fr/");
      expect(hauteSavoieLink).toHaveAttribute("target", "_blank");
      expect(hauteSavoieLink).toHaveAttribute("rel", "noopener noreferrer");
    });
  });

  describe("Main footer section", () => {
    it("should render the CPMB logo and description", () => {
      render(<Footer />);

      const logo = screen.getByAltText("Logo Chœur des Pays du Mont-Blanc");
      expect(logo).toBeInTheDocument();
      expect(logo).toHaveAttribute("src", "/CPMB-logo-blanc.png");
      expect(logo).toHaveAttribute("width", "160");
      expect(logo).toHaveAttribute("height", "55");

      const description = screen.getByText("Partager la passion de la musique chorale au cœur des Alpes.");
      expect(description).toBeInTheDocument();
    });

    it("should render the partners section with correct links", () => {
      render(<Footer />);

      expect(screen.getByRole("heading", { name: "Nos Partenaires" })).toBeInTheDocument();

      // Partner links in the text list
      const veranTextLink = screen.getByRole("link", { name: "Veran Piano" });
      expect(veranTextLink).toHaveAttribute("href", "https://www.veran-piano.com/");
      expect(veranTextLink).toHaveAttribute("target", "_blank");
      expect(veranTextLink).toHaveAttribute("rel", "noopener noreferrer");

      const gaillardTextLink = screen.getByRole("link", { name: "Ville de Gaillard" });
      expect(gaillardTextLink).toHaveAttribute("href", "https://www.gaillard.fr/");
      expect(gaillardTextLink).toHaveAttribute("target", "_blank");
      expect(gaillardTextLink).toHaveAttribute("rel", "noopener noreferrer");

      const hauteSavoieTextLink = screen.getByRole("link", { name: "Département de la Haute-Savoie" });
      expect(hauteSavoieTextLink).toHaveAttribute("href", "https://www.hautesavoie.fr/");
      expect(hauteSavoieTextLink).toHaveAttribute("target", "_blank");
      expect(hauteSavoieTextLink).toHaveAttribute("rel", "noopener noreferrer");
    });

    it("should render the site map section with internal links", () => {
      render(<Footer />);

      expect(screen.getByRole("heading", { name: "Plan du site" })).toBeInTheDocument();

      const homeLink = screen.getByRole("link", { name: "Page d'accueil" });
      expect(homeLink).toHaveAttribute("href", "/");

      const legalLink = screen.getByRole("link", { name: "Mentions légales" });
      expect(legalLink).toHaveAttribute("href", "/mentions-legales");
    });

    it("should render the copyright notice with current year", () => {
      render(<Footer />);

      const copyright = screen.getByText("© 2024 Chœur des Pays du Mont-Blanc. Tous droits réservés.");
      expect(copyright).toBeInTheDocument();
    });
  });

  describe("Accessibility", () => {
    it("should have proper semantic structure", () => {
      render(<Footer />);

      const footer = screen.getByRole("contentinfo");
      expect(footer).toBeInTheDocument();
      expect(footer.tagName).toBe("FOOTER");
    });

    it("should have proper heading hierarchy", () => {
      render(<Footer />);

      const headings = screen.getAllByRole("heading");
      expect(headings).toHaveLength(2);

      const partnersHeading = screen.getByRole("heading", { name: "Nos Partenaires" });
      expect(partnersHeading.tagName).toBe("H4");

      const siteMapHeading = screen.getByRole("heading", { name: "Plan du site" });
      expect(siteMapHeading.tagName).toBe("H4");
    });

    it("should have proper ARIA labels for external links", () => {
      render(<Footer />);

      // Check that all external links have proper aria-labels
      const externalLinksWithLabels = [
        "Visitez le site de Veran Piano - nouvelle fenêtre",
        "Visitez le site de la Ville de Gaillard - nouvelle fenêtre",
        "Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre",
      ];

      externalLinksWithLabels.forEach((label) => {
        expect(screen.getByLabelText(label)).toBeInTheDocument();
      });
    });

    it("should have proper security attributes for external links", () => {
      render(<Footer />);

      const externalLinks = screen.getAllByRole("link").filter((link) => link.getAttribute("href")?.startsWith("http"));

      externalLinks.forEach((link) => {
        expect(link).toHaveAttribute("target", "_blank");
        expect(link).toHaveAttribute("rel", "noopener noreferrer");
      });
    });
  });

  describe("Responsive design classes", () => {
    it("should have responsive layout classes", () => {
      render(<Footer />);

      const footer = screen.getByRole("contentinfo");
      expect(footer).toHaveClass("font-noto");

      // Check for responsive flex classes in main content
      const mainContent = footer.querySelector(
        ".container.mx-auto.flex.flex-col.items-start.gap-4.py-4.lg\\:flex-row.lg\\:gap-24"
      );
      expect(mainContent).toBeInTheDocument();
    });
  });

  describe("Content validation", () => {
    it("should contain all expected text content", () => {
      render(<Footer />);

      // Main description
      expect(screen.getByText("Partager la passion de la musique chorale au cœur des Alpes.")).toBeInTheDocument();

      // Section headings
      expect(screen.getByText("Nos Partenaires")).toBeInTheDocument();
      expect(screen.getByText("Plan du site")).toBeInTheDocument();

      // Partner names
      expect(screen.getByText("Veran Piano")).toBeInTheDocument();
      expect(screen.getByText("Ville de Gaillard")).toBeInTheDocument();
      expect(screen.getByText("Département de la Haute-Savoie")).toBeInTheDocument();

      // Site map links
      expect(screen.getByText("Page d'accueil")).toBeInTheDocument();
      expect(screen.getByText("Mentions légales")).toBeInTheDocument();
    });
  });
});
