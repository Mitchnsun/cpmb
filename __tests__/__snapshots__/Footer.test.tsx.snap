// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should render the footer element 1`] = `
<div>
  <footer
    class="font-noto"
  >
    <section
      class="container mx-auto flex items-center justify-around gap-6 py-4 text-center"
    >
      <a
        aria-label="Visitez le site de Veran Piano - nouvelle fenêtre"
        class="basis-1/3"
        href="https://www.veran-piano.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="Logo Veran Piano"
          class="m-auto"
          height="120"
          src="/logo-veran-pianos.png"
          width="200"
        />
      </a>
      <a
        aria-label="Visitez le site de la Ville de Gaillard - nouvelle fenêtre"
        class="basis-1/3"
        href="https://www.gaillard.fr/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="Logo Ville de Gaillard"
          class="m-auto"
          height="120"
          src="/logo-gaillard.png"
          width="200"
        />
      </a>
      <a
        aria-label="Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre"
        class="basis-1/3"
        href="https://www.hautesavoie.fr/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="Logo Haute-Savoie"
          class="m-auto"
          height="55"
          src="/haute-savoie.svg"
          width="90"
        />
      </a>
    </section>
    <div
      class="bg-gray-800 px-4"
    >
      <div
        class="container mx-auto flex flex-col items-start gap-4 py-4 lg:flex-row lg:gap-24"
      >
        <div>
          <img
            alt="Logo Chœur des Pays du Mont-Blanc"
            class="shrink-0"
            height="55"
            src="/CPMB-logo-blanc.png"
            width="160"
          />
          <p
            class="mt-4 text-sm text-gray-400"
          >
            Partager la passion de la musique chorale au cœur des Alpes.
          </p>
        </div>
        <section>
          <h4
            class="font-bold tracking-tight text-lg mb-2 text-white"
          >
            Nos Partenaires
          </h4>
          <ul
            class="space-y-1 text-sm"
          >
            <li>
              <a
                aria-label="Visitez le site de Veran Piano - nouvelle fenêtre"
                class="text-gray-400 hover:text-white"
                href="https://www.veran-piano.com/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Veran Piano
              </a>
            </li>
            <li>
              <a
                aria-label="Visitez le site de la Ville de Gaillard - nouvelle fenêtre"
                class="text-gray-400 hover:text-white"
                href="https://www.gaillard.fr/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Ville de Gaillard
              </a>
            </li>
            <li>
              <a
                aria-label="Visitez le site du Département de la Haute-Savoie - nouvelle fenêtre"
                class="text-gray-400 hover:text-white"
                href="https://www.hautesavoie.fr/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Département de la Haute-Savoie
              </a>
            </li>
          </ul>
        </section>
        <section>
          <h4
            class="font-bold tracking-tight text-lg mb-2 text-white"
          >
            Plan du site
          </h4>
          <ul
            class="space-y-1 text-sm"
          >
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/"
              >
                Page d'accueil
              </a>
            </li>
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/presentation"
              >
                Présentation
              </a>
            </li>
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/presse"
              >
                Presse
              </a>
            </li>
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/mentions-legales"
              >
                Mentions légales
              </a>
            </li>
          </ul>
        </section>
        <section>
          <h4
            class="font-bold tracking-tight text-lg mb-2 text-white"
          >
            Artistes
          </h4>
          <ul
            class="space-y-1 text-sm"
          >
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/presentation/benoit-dubu"
              >
                Benoît Dubu
              </a>
            </li>
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/presentation/agnes-lorincz"
              >
                Ágnes Lőrincz
              </a>
            </li>
            <li>
              <a
                class="text-gray-400 hover:text-white"
                href="/presentation/marjorie-saunier"
              >
                Marjorie Saunier
              </a>
            </li>
          </ul>
        </section>
      </div>
      <p
        class="container mx-auto border-t border-gray-700 py-4 text-center text-sm text-gray-500"
      >
        © 
        2024
         Chœur des Pays du Mont-Blanc. Tous droits réservés - Conçu et développé par 
        <a
          aria-label="Visitez le site de Go Cosmic - nouvelle fenêtre"
          class="text-sky-600 hover:underline"
          href="https://www.gocosmic.dev/fr"
          rel="noopener noreferrer"
          target="_blank"
        >
          Go Cosmic
        </a>
      </p>
    </div>
  </footer>
</div>
`;
